pytket/phir/phirgen_parallel.py

```diff
@@ -0,0 +1,348 @@
+        # this means we can treat all other angled gates the same
+        else:
+            fmt_g2q: dict[int, list[tk.Command]] = {0: []}
+            for shard in group:
+                fmt_g2q[0].append(shard.primary_command)
+            groups2qops(fmt_g2q, ops)
+
+
+def genphir_parallel(
+    inp: list[tuple[Ordering, ShardLayer, Cost]], machine: Machine
+) -> str:
+    """Convert a list of shards to the equivalent PHIR with parallel gating.
+
+    Args:
+        inp: list of shards
+        machine: a QTM machine on which to simulate the circuit
+    """
+    max_parallel_tq_gates = len(machine.tq_options)
+    max_parallel_sq_gates = len(machine.sq_options)
+
+    phir: dict[str, Any] = {
+        "format": "PHIR/JSON",
+        "version": "0.1.0",
+        "metadata": {"source": "pytket-phir", "strict_parallelism": "true"},
+    }
+    ops: list[dict[str, Any]] = []
+
+    qbits = set()
+    cbits = set()
+    for _orders, shard_layer, layer_cost in inp:
+        # within each shard layer, create groups of parallelizable shards
+        # squash all the sub-commands into the first shard in the group
+        shard_groups = process_shards(
+            shard_layer, max_parallel_tq_gates, max_parallel_sq_gates
+        )
+        for group_number in sorted(shard_groups.keys()):
+            group = shard_groups[group_number]
+            for shard in group:
+                qbits |= shard.qubits_used
+                cbits |= shard.bits_read | shard.bits_written
+                if shard.sub_commands.values():
+                    # sub-commands are always sq gates
+                    subcmd_groups = process_sub_commands(
+                        shard.sub_commands, max_parallel_sq_gates
+                    )
+                    groups2qops(subcmd_groups, ops)
+            format_and_add_primary_commands(group, ops)
+
+        ops.append(
+            {
+                "mop": "Transport",
+                "duration": (layer_cost, "ms"),
+            },
+        )
+
+    # TODO(kartik): this may not always be accurate
+    # https://github.com/CQCL/pytket-phir/issues/24
+    qvar_dim: dict[str, int] = {}
+    for qbit in qbits:
+        qvar_dim.setdefault(qbit.reg_name, 0)
+        qvar_dim[qbit.reg_name] += 1
+
+    cvar_dim: dict[str, int] = {}
+    for cbit in cbits:
+        cvar_dim.setdefault(cbit.reg_name, 0)
+        cvar_dim[cbit.reg_name] += 1
+
+    decls: list[dict[str, str | int]] = [
+        {
+            "data": "qvar_define",
+            "data_type": "qubits",
+            "variable": qvar,
+            "size": dim,
+        }
+        for qvar, dim in qvar_dim.items()
+    ]
+
+    decls += [
+        {
+            "data": "cvar_define",
+            "data_type": "u32",
+            "variable": cvar,
+            "size": dim,
+        }
+        for cvar, dim in cvar_dim.items()
+    ]
+
+    phir["ops"] = decls + ops
+    PHIRModel.model_validate(phir)
+    return json.dumps(phir)
```

Comment:
Another option, since parallel phirgen is becoming the default when a machine arg is provided, is to consolidate the code in a single file to handle both scenarios.

Reply:
I think if we put it all in one file, it would just end up being really long and we would move a bunch of the functions out into a separate utils file anyway, I think if we are going to need 2 files anyway, we may as well keep the division as parallel/serial rather than gen/utils